### PR TITLE
Major redesign

### PR DESCRIPTION
**Pull request by poelzi**
These are the collective changes we did for getting a working collection of metrics from multiple servers.
Sorry for the patchbomb, but it had to go fast.

- Implement multiple new collectors for ethernet port statistics. RX/TX/port stats
- All data collection is now done in a seperate thread and only the last results are reported back on request.
We are already over 3.3 mb of metrics and run into multiple timeouts from telegraf and very unstable behaviour.
Now the socket pushes out over 1.1 gb/s in metrics.
The current design could be done a bit better with job queues but this can be done later
- Better handling of failures through wrapping of query methods
- Generic query class for gauge metrics
- Proper fix for illegal password characters 
- setup.py for installation
- logging framework with debug levels instead of printing

### CODE REVIEW

**Comment by mithunmgore** [comment]
@poelzi I liked it 👍 I will test.

File: modules/BaseCollector.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        for retry in range(2):
+            try:
+                return fnc(*tuple(args), **dict(kwargs))
+                # yield data
+            except urllib.error.URLError as e:
+                logger.error("URLError: ", server, e.reason)
+            except UcsException as e:
+                if e.error_code == 552:
+                    # we need to login again
+                    handle.logout()
```
Comment (mithunmgore) [inline modules/BaseCollector.py]: wont get handle

Comment (poelzi) [reply modules/BaseCollector.py]: fixed

**Comment by poelzi** [comment]
I also changed the way metrics are reported in accordance with our discussion of the server monitoring team.
Metrics are only reported once after collection.

**Comment by poelzi** [comment]
@mithunmgore ping

File: ucs-exporter.py [hunk 2]
```diff
@@ -42,26 +54,44 @@ def get_params():
...
 
+    return manager
 
 if __name__ == '__main__':
     params = get_params()
-    register_collectors(params)
-    start_http_server(9876)
-    while True:
-        time.sleep(10)
+    print(params)
```
Comment (viennaa) [inline ucs-exporter.py:92]: Either move to logging framework or remove

File: modules/ConnectionManager.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    def update_state(self):
+        """
+        Return handles to all the servers from inventory.
+        :return: dict of handles
+                e.g. self.handles = {"server_name": "<server_handle>", ...}
+        """
+        server_list = self.get_inventory()
+        logger.debug("Create handles for Ucsm Servers: %s", server_list)
+        #import traceback
+        #traceback.print_stack()
```
Comment (viennaa) [inline modules/ConnectionManager.py:136]: Remove

File: modules/ConnectionManager.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        :return: list of servers
+        """
+        config = self.get_config_data()
+        server_list = []
+        if 'netbox' in config and len(config['netbox']):
+            from modules.Netbox import Netbox
+
+            netbox_data = config['netbox']
+            netbox_obj = Netbox(nb_config=netbox_data)
+            server_list += netbox_obj.get_ucsm_servers_from_regions(netbox_data["regions"])
```
Comment (viennaa) [inline modules/ConnectionManager.py:64]: This should be try-excepted

File: modules/ConnectionManager.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        if server not in self.handles:
+            logger.debug("Login into %s" % server)
+            srv_obj = UcsmServer(server, self.creds['username'], self.creds['master_password'], self.config['domain'])
+            if not srv_obj.handle:
+                if self.config['retry_timeout'] > 0:
+                    self.blacklist[server] = time.time() + self.config['retry_timeout']
+                    logger.error("Problem creating login to %s. Retry in %s seconds",
+                                server, self.config['retry_timeout'], extra={'server': server})
+                else:
+                    logger.error("Problem creating login to %s", server, extra={'server': server})
```
Comment (viennaa) [inline modules/ConnectionManager.py:115]: Don't we want to return in both cases? I will always run into `handle.login` exception. I mean it is catched, but can be avoided.

File: modules/ConnectionManager.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+    def get_collectors(self):
+        return self._collectors
+
+    def start_poll_thread(self, host):
+        current = self._poll_threads.get(host, None)
+        if current:
+            if not current.is_alive():
+                # thread is dead, create a new one
+                self._poll_threads.delete(host)
```
Comment (viennaa) [inline modules/ConnectionManager.py:200]: Have you considered using a queue to check for existing threads or emptiness?

**Review by viennaa — COMMENTED** [review]
I like the approach. Threading the whole thing out indeed improves the throughput a lot. 

Just a few annotations from my side inline.

File: modules/ConnectionManager.py (hunk 3 above)
Comment (mithunmgore) [reply modules/ConnectionManager.py:136]: Removed

File: ucs-exporter.py (hunk 2 above)
Comment (mithunmgore) [reply ucs-exporter.py:92]: Removed

File: modules/ConnectionManager.py (hunk 4 above)
Comment (mithunmgore) [reply modules/ConnectionManager.py:64]: Done

File: modules/ConnectionManager.py (hunk 5 above)
Comment (mithunmgore) [reply modules/ConnectionManager.py:115]: Returning value.

File: modules/ConnectionManager.py (hunk 6 above)
Comment (mithunmgore) [reply modules/ConnectionManager.py:200]: I am working on adding queue. will create another PR.

**Comment by mithunmgore** [comment]
@viennaa Please check the PR https://github.com/sapcc/ucs-exporter/pull/9/ for addressed comments and new collectors

**Pull request merged**